palace/fem/mesh.cpp

```diff
@@ -0,0 +1,369 @@
+{
+  // Count the number of elements of each type in the local mesh.
+  std::unordered_map<mfem::Geometry::Type, int> counts;
+  for (int i = start; i < stop; i++)
+  {
+    const auto geom = use_bdr ? mesh.GetBdrElementGeometry(i) : mesh.GetElementGeometry(i);
+    auto it = counts.find(geom);
+    if (it == counts.end())
+    {
+      counts[geom] = 1;
+    }
+    else
+    {
+      it->second++;
+    }
+  }
+
+  // Populate the indices arrays for each element geometry.
+  std::unordered_map<mfem::Geometry::Type, int> offsets;
+  std::unordered_map<mfem::Geometry::Type, std::vector<int>> element_indices;
+  for (auto it = counts.begin(); it != counts.end(); ++it)
+  {
+    offsets[it->first] = 0;
+    element_indices[it->first] = std::vector<int>(it->second);
+  }
+  for (int i = start; i < stop; i++)
+  {
+    const auto geom = use_bdr ? mesh.GetBdrElementGeometry(i) : mesh.GetElementGeometry(i);
+    auto &offset = offsets[geom];
+    auto &indices = element_indices[geom];
+    indices[offset++] = i;
+  }
+
+  return element_indices;
+}
+
+template <typename T>
+auto AssembleGeometryData(const mfem::GridFunction &mesh_nodes, Ceed ceed,
+                          mfem::Geometry::Type geom, std::vector<int> &indices,
+                          T GetCeedAttribute)
+{
+  const mfem::FiniteElementSpace &mesh_fespace = *mesh_nodes.FESpace();
+  const mfem::Mesh &mesh = *mesh_fespace.GetMesh();
+
+  ceed::CeedGeomFactorData data;
+  data.dim = mfem::Geometry::Dimension[geom];
+  data.space_dim = mesh.SpaceDimension();
+  data.indices = std::move(indices);
+  const std::size_t num_elem = data.indices.size();
+
+  // Allocate storage for geometry factor data (stored as attribute + quadrature weight +
+  // Jacobian).
+  CeedElemRestriction mesh_restr =
+      FiniteElementSpace::BuildCeedElemRestriction(mesh_fespace, ceed, geom, data.indices);
+  CeedBasis mesh_basis = FiniteElementSpace::BuildCeedBasis(mesh_fespace, ceed, geom);
+  CeedInt num_qpts, geom_data_size = 2 + data.space_dim * data.dim;
+  PalaceCeedCall(ceed, CeedBasisGetNumQuadraturePoints(mesh_basis, &num_qpts));
+  PalaceCeedCall(
+      ceed, CeedVectorCreate(ceed, num_elem * num_qpts * geom_data_size, &data.geom_data));
+
+  // Data for quadrature point i, component j, element k is found at index i * strides[0] +
+  // j * strides[1] + k * strides[2].
+  CeedMemType mem;
+  CeedInt strides[3];
+  PalaceCeedCall(ceed, CeedGetPreferredMemType(ceed, &mem));
+  if (mfem::Device::Allows(mfem::Backend::DEVICE_MASK) && mem == CEED_MEM_DEVICE)
+  {
+    // GPU backends have CEED_STRIDES_BACKEND = {1, num_elem * num_qpts, num_qpts}.
+    strides[0] = 1;
+    strides[1] = num_elem * num_qpts;
+    strides[2] = num_qpts;
+  }
+  else
+  {
+    // CPU backends have CEED_STRIDES_BACKEND = {1, num_qpts, num_qpts * geom_data_size}.
+    strides[0] = 1;
+    strides[1] = num_qpts;
+    strides[2] = num_qpts * geom_data_size;
+  }
```

Comment:
libCEED uses different "prefered" backend layouts for CPU and GPU backend E-vectors. Ideally this would be handled by just prescribing `CEED_STRIDES_BACKEND` but unfortunately I still need to know the strides in order to access various entries of the quadrature data vector from outside of libCEED. So I tried to choose these so that the strided element restriction is as efficient as possible.